Clean up studio edit template

mtyaka opened the pull request:
**Description**:

This PR cleans up the Studio edit template and associated JS code to:

- Use `field.display_name` values for field labels instead of hardcoding/duplicating the names in the template.
- Directly display help text under associated input field instead of `title` and `<span class="sr">`.
- Replace usage of HTML `id` attributes with `class`-es. In cases where that is not possible, ensure that the `id` attributes are globally unique.
- Use `<button>` elements instead of `<a>` elements for buttons.
- Replace input placeholders with help texts visible to everyone.

**Screenshots**:

Before:

<img width="960" alt="screen shot 2016-08-04 at 17 53 48" src="https://cloud.githubusercontent.com/assets/32585/17397985/1c4c2a36-5a6d-11e6-9521-86a0551a1fc3.png">

After:

<img width="965" alt="screen shot 2016-08-04 at 17 52 52" src="https://cloud.githubusercontent.com/assets/32585/17397996/246f5558-5a6d-11e6-94c4-94e4c2bac7ca.png">

**JIRA ticket**: https://openedx.atlassian.net/browse/SOL-1990

**Dependencies**: None

**Discussions**: https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/77#discussion_r71034788

**Sandbox URLs** (installed version 9845083078d129009933cd8c8106b27c14e17242):

- LMS: http://dndv2-sandbox5.opencraft.hosting/
- Studio: http://studio-dndv2-sandbox5.opencraft.hosting/

You are probably mostly interested in the Studio edit functionality. Some DnDv2 blocks can be edited at: http://studio-dndv2-sandbox5.opencraft.hosting/container/block-v1:OpenCraftX+OC1+1+type@vertical+block@e3c62b19090948b8ad89d3386bff4435

**Testing Instructions**:

This PR does not implement any new functionality, so there is no specific feature to test.
The important thing is to ensure that the studio edit functionality continues to look and function correctly.

**Reviewers**:

- [x] OpenCraft: @bradenmacdonald
- [x] TNL: @cahrens and/or @staubina
- [ ] ~~a11y: @edx/edx-accessibility~~ Scheduled combined review for all DnDv2.1 stories: [SOL-1978](https://openedx.atlassian.net/browse/SOL-1978)
- [x] Product: @sstack22

edx-webhook commented:
Thanks for the pull request, @mtyaka! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx-solutions%2Fxblock-drag-and-drop-v2&number=88

bradenmacdonald commented on drag_and_drop_v2/public/css/drag_and_drop_edit.css:
```diff
@@ -222,6 +242,7 @@
     display: none;
 }
 .xblock--drag-and-drop--editor .items-form .row.advanced-link {
+    cursor: pointer;
```
This is because the `<a>` element has no `href=""` attribute, right? Is it better to do this or use `href="#"` + `preventDefault()` ?

mtyaka replied:
Yes, that's right. I removed `href="#"` and added this CSS rule. I believe this is considered to be a better approach than `href="#"`, which has always felt like a hack to me.

cahrens replied:
Instead of using an a-tag, you should use a button. A-tag elements should take you someplace, whereas buttons do actions. Buttons give you the correct a11y behavior, whereas you have to manually create those behaviors with other elements using aria properties.

Tabbing through the Studio editing UI, I don't see any way to even get to these "Remove" links with the keyboard.

mtyaka replied:
@cahrens Thanks, thanks makes sense. I replaced `<a>` elements with `<button>`s.

bradenmacdonald commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -227,6 +234,7 @@ def studio_view(self, context):
             'js_templates': js_templates,
             'help_texts': help_texts,
             'field_values': field_values,
```
It seems inconsistent to pass the help text and field values as context vars, but to use `self.fields.foo.display_name` to get the display name. Maybe we should just pass `fields` as a context variable and then use `fields.foo.display_name`, `fields.food.help`, and `fields.foo.values`? That would be more consistent and efficient.

mtyaka replied:
Good idea, passing `fields` only makes it more consistent; done.

bradenmacdonald commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -117,20 +149,24 @@
+                    <label class="h3">
+                        <span>{% trans self.fields.item_background_color.display_name %}</span>
+                        <input class="item-background-color"
+                               placeholder="e.g. blue or #0000ff"
```
Needs i18n. `blue` is a CSS keyword so shouldn't be translated, but "e.g." is English-specific. So this should be marked for i18n as

```
{% blocktrans with example1="blue" example2="#0000ff" %}
e.g. {{example1}} or {{example2}}
{% endblocktrans %}
```

bradenmacdonald commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -117,20 +149,24 @@
+                    <label class="h3">
+                        <span>{% trans self.fields.item_text_color.display_name %}</span>
+                        <input class="item-text-color"
+                               placeholder="e.g. white or #ffffff"
```
Also needs i18n

bradenmacdonald commented on drag_and_drop_v2/templates/html/js_templates.html:
```diff
@@ -9,80 +9,84 @@
-                   class="size width"
-                   value="{{ zone.width }}" />
-            <label for="zone-{{index}}-height">{{i18n "height"}}</label>
+                   class="title"
```
This class name (`title`) is too generic; Studio already defines a `[class*="view-"] .modal-window .title` rule which is affecting this. I'd suggest `zone-title` instead.

bradenmacdonald commented on drag_and_drop_v2/templates/html/js_templates.html:
```diff
@@ -9,80 +9,84 @@
-        <div class="layout">
-            <label for="zone-{{index}}-width">{{i18n "width"}}</label>
+        <label>
+            <span>{{i18n "Text"}}</span>
```
Maybe we should change this to "Title" ?

bradenmacdonald commented on drag_and_drop_v2/templates/html/js_templates.html:
```diff
@@ -9,80 +9,84 @@
-                   id="zone-{{index}}-height"
-                   class="size height"
-                   value="{{ zone.height }}" />
+                   class="description"
```
This class name is also too generic and likely to be affected by external CSS rules.

bradenmacdonald commented:
@mtyaka Just a few minor notes/questions - thanks for this!

mtyaka commented:
@bradenmacdonald Thanks for the review! Your comments have been addressed.

bradenmacdonald commented:
👍  Nice improvement, @mtyaka.

- [x] I tested this:
   * Made changes to the dnd components on the Sandbox, using the updated Studio editor.
   * I used `?preview-lang=eo` on the sandbox to test some i18n, though the edx-platform is currently only able to translate strings that already exist in the platform and does not yet always load xblock-specific strings.
   * I used chrome's element inspector to copy the rendered outerHTML for each tab of the Studio editor and search for `id="..."` attributes that had non-unique values.
- [x] I read through the code
- [x] Includes documentation: no readme changes required. This PR adds brief documentation where it belongs - inline where the user is making changes.

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -215,18 +216,16 @@ def studio_view(self, context):
+        # of the DnDv2 block on the same page sharing the same ID values.
+        # We avoid using ID attributes in preference to classes, but sometimes we still need IDs to
+        # connect 'for' and 'aria-describedby' attributes to the associated elements.
+        id_suffix = uuid.uuid4().hex[:16]
```
Another way of creating unique IDs is to incorporate the XBlock id. That's what we do for capa and ORA.

mtyaka replied:
@cahrens Thanks for the hint. I replaced `uuid.uuid4()` with `block.location.html_id()`. I was not aware of the `html_id()` function before.

cahrens commented:
I would suggest a doc review of the field help text; it's pretty inconsistent in formatting and style. Who is the edX doc person working on Drag and Drop?

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -215,18 +216,16 @@ def studio_view(self, context):
+        # connect 'for' and 'aria-describedby' attributes to the associated elements.
+        id_suffix = uuid.uuid4().hex[:16]
+        js_templates = js_templates.replace('{{id_suffix}}', id_suffix)
         context = {
```
Why are you directly replacing id_suffix on line 224 in js_templates (I'm not sure how those are used) and also passing id_suffix in the context (which is what I would expect)?

mtyaka replied:
@cahrens js_templates contains Handlebars templates. Replacing `id_suffix` with raw string substitutions is a bit of a hack, however I think it's still better than passing the `id_suffix` variable to JavaScript and making sure it is present in the handlebars context every time when rendering a handlebars template.

We still have to pass the `id_suffix` variable in the context, because we use it inside django templates, too.

bradenmacdonald commented:
^ @pdesjardins is doing docs for drag and drop

cahrens commented on drag_and_drop_v2/public/css/drag_and_drop_edit.css:
```diff
@@ -254,14 +275,15 @@
 }
 
 .xblock--drag-and-drop--editor .remove-zone {
+    cursor: pointer;
```
Same note here-- don't use a-tags.

cahrens commented on drag_and_drop_v2/public/css/drag_and_drop_edit.css:
```diff
@@ -254,14 +275,15 @@
 }
 
 .xblock--drag-and-drop--editor .remove-item {
+    cursor: pointer;
```
Same note here-- don't use a-tags.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -67,15 +89,21 @@
+                    <input class="background-url"
+                           id="background-url-{{id_suffix}}"
                            type="text"
                            placeholder="{% trans 'For example, http://example.com/background.png or /static/background.png' %}">
```
Placeholders should be avoided-- they are too low in contrast, and they don't provide a very good user experience because the text disappears when you start typing. It's better to just add a visible (to all) description.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -87,13 +115,17 @@
-                    <label for="display-labels">{% trans "Display label names on the image" %}:</label>
-                    <input name="display-labels" id="display-labels" type="checkbox" />
+                    <label>
+                        <span>{% trans "Display label names on the image" %}:</span>
```
Include the colon in the text to be translated-- colons won't make sense in all languages. Or is it more natural to have the checkbox before the text, so there is no need for a colon?

mtyaka replied:
@cahrens I think having the checkbox before the text is more natural, so I swapped them and removed the colon.

cahrens commented:
I imagine these font sizes (12?) are too small. Hopefully an a11y review can weigh in.

![image](https://cloud.githubusercontent.com/assets/484484/17500995/b05c28e6-5da8-11e6-8bd2-2e74eca13d3c.png)

cahrens commented:
There are color contrast errors in this "blue" section.

![image](https://cloud.githubusercontent.com/assets/484484/17501047/0dd02f68-5da9-11e6-8407-ca53b9a6e2e1.png)

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -117,21 +149,25 @@
+                    <label class="h3">
+                        <span>{% trans fields.item_background_color.display_name %}</span>
+                        <input class="item-background-color"
+                               placeholder="{% blocktrans with example1='blue' example2='#0000ff' %}e.g. {{example1}} or {{example2}}{% endblocktrans %}"
```
Again, avoid placeholders.

cahrens commented on drag_and_drop_v2/templates/html/js_templates.html:
```diff
@@ -9,80 +9,84 @@
-                   value="{{ zone.height }}" />
+                   class="zone-description"
+                   value="{{ zone.description }}"
+                   placeholder="{{i18n 'Describe this zone to non-visual users'}}"
```
Avoid placeholders.

cahrens commented:
There are some good improvements here! However, there is still more work to do to make the Studio UI accessible.

I'm sure a full a11y review will find things I have missed, but for now, try accessing everything in the UI just with just the keyboard-- there are several elements that one cannot currently navigate to (specifically, there are many problematic a-tags that are acting as buttons).

staubina commented:
I will wait to review in depth until updates are made per @cahrens comments

staubina commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -87,13 +115,17 @@
-                    <label for="display-borders">{% trans "Display zone borders on the image" %}:</label>
-                    <input name="display-borders" id="display-borders" type="checkbox" />
+                    <label>
+                        <span>{% trans "Display zone borders on the image" %}:</span>
```
Another color take a  look at @cahrens comment above

staubina commented on drag_and_drop_v2/templates/html/js_templates.html:
```diff
@@ -126,32 +130,38 @@
             <label>
-                {{i18n "Preferred width as a percentage of the background image width (or blank for automatic width):"}}
+                <span>
+                  {{i18n "Preferred width as a percentage of the background image width (or blank for automatic width):"}}
```
Should we include this colon in the translation?

mtyaka replied:
@staubina I changed the way this field is displayed in the last commit to make it more consistent with other fields. The new layout does not need the colon, so I removed it.

staubina commented:
There seems to be a lack of consistency in the font sizes throughout the code. Is this something that could be updated here or in a later PR?

![screen shot 2016-08-09 at 8 27 33 am](https://cloud.githubusercontent.com/assets/2854941/17516249/3c5fb12a-5e0b-11e6-824c-825df7f2c250.png)

mtyaka commented:
Thanks for the review @cahrens and @staubina!

I pushed an updated that addresses your comments, including making font sizes more consistent. I updated the sandbox with the latest code, please take another look.

Here are some screenshots:

<img width="993" alt="screen shot 2016-08-10 at 13 44 35" src="https://cloud.githubusercontent.com/assets/32585/17543703/16a38be6-5f05-11e6-8448-d2391121b690.png">

<img width="992" alt="screen shot 2016-08-10 at 13 45 06" src="https://cloud.githubusercontent.com/assets/32585/17543702/15fcd238-5f05-11e6-97e1-2c7bbf0fc71a.png">

<img width="991" alt="screen shot 2016-08-10 at 13 45 34" src="https://cloud.githubusercontent.com/assets/32585/17543709/21783fb2-5f05-11e6-9321-40ed7a906df1.png">

staubina commented:
👍

cahrens commented:
This is looking a lot better! Make sure that when a button is clicked, the focus goes someplace sensible. For instance, when clicking "Add an Item", focus should move to the newly created Item.

I also find it odd that when focus is on a button (both "Add an Item" and the Delete buttons), they actually look disabled.

![image](https://cloud.githubusercontent.com/assets/484484/17570046/d0a48d7a-5f18-11e6-8ac0-4fbe81c4a918.png)

cahrens commented:
Perhaps beyond the scope of this PR, but note that the Continue and Cancel buttons at the bottom of the dialog continue to be a-links. I noticed it because I hit Space on Continue (a button is triggered by both Return and Space), and I was confused because the page scrolled.

cahrens commented:
Have you tested for RTL? You can do that (on your sandbox) by adding ?preview-lang=rtl to the end of the URL. I see some minor layout issues.

If you'd like a review of the CSS, I can ask the FED person on my team to review. I'm not nearly as good with CSS, and also I'm not sure what sort of standards we try to maintain for code outside of edx-platform (that isn't using the pattern library).

cahrens commented on drag_and_drop_v2/public/css/drag_and_drop_edit.css:
```diff
@@ -208,22 +256,27 @@
 }
 
 .xblock--drag-and-drop--editor .items-form .zone-checkbox-label {
     text-align: left;
```
Things like this are likely problematic for RTL.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -86,21 +117,29 @@
                         <form class="zones-form"></form>
-                        <a href="#" class="add-zone add-element"><div class="icon add"></div>{% trans "Add a zone" %}</a>
+                        <button class="btn add-zone add-element">
+                            <span class="icon add"></span>
```
I'm not sure what type of icon this is, but for Font Awesome icons we always specify aria-hidden (with either visible text like this or screen reader text).

See example HTML here--
http://ux.edx.org/design_elements/icons/

mtyaka commented:
Thanks @cahrens, I addressed most of your comments:

- Focus goes to the new item when clicking the "Add Zone" or "Add Item" button.
- Fix `:hover` and `:focus` styles so that buttons don't look disabled.
- Fixed several RTL layout issues.
- Added `aria-hidden="true"` to icons.

> Perhaps beyond the scope of this PR, but note that the Continue and Cancel buttons at the bottom of the dialog continue to be a-tags.

That button HTML code was copied from the Studio (CAPA problems currently also use `<a>` tags) so that we can use default Studio styles and avoid having to copy/imitate the look using our own CSS. I tried switching to button elements, but that broke the styles, so I would prefer to wait until `<a>` are replaced with `<buttons>` in the Studio, after which we can simply copy the new HTML structure.

> If you'd like a review of the CSS, I can ask the FED person on my team to review.

A review of the CSS would be great, but I think it would be best if it was done outside of this PR (reviewing all CSS, not just the changes in this PR).

cahrens commented:
I understand about the Continue/Cancel buttons. When Studio was written we incorrectly used a-tags everywhere, and that is something that will be cleaned up later this year.

However, when I tab to the Continue button and press Enter, focus is remaining in the button bar-- I think this is new behavior since yesterday. After Continue, keyboard focus needs to move to the top of the new settings panel. As it is, it is very easy to move keyboard focus outside of the dialog entirely (while tabbing to try to figure out where focus is), and difficult to figure out how to get back.

Also a nit about adding a little padding around this close button:

![image](https://cloud.githubusercontent.com/assets/484484/17587126/39867ff8-5f94-11e6-911a-9108ff5ebd6c.png)

mtyaka commented:
@cahrens 

> However, when I tab to the Continue button and press Enter, focus is remaining in the button bar-- I think this is new behavior since yesterday. After Continue, keyboard focus needs to move to the top of the new settings panel. As it is, it is very easy to move keyboard focus outside of the dialog entirely (while tabbing to try to figure out where focus is), and difficult to figure out how to get back.

The behavior was not new, but I should have realized and fixed it after your previous comment, sorry I missed that. It's fixed in the latest update.

> Also a nit about adding a little padding around this close button.

Last commit adds a bit of extra padding on small screens.

I updated the sandbox with latest changes.

staubina commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -90,22 +90,22 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
     item_background_color = String(
         display_name=_("Item background color"),
-        help=_("The background color of draggable items in the problem."),
+        help=_("The background color of draggable items in the problem (e.g. 'blue' or '#0000ff')."),
```
Should we use "example" instead of e.g.? I am not sure how well e.g. translates.

staubina commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -90,22 +90,22 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
     item_text_color = String(
         display_name=_("Item text color"),
-        help=_("Text color to use for draggable items."),
+        help=_("Text color to use for draggable items (e.g. 'white' or '#ffffff')."),
```
Same as last comment on the use of e.g.

staubina commented:
The rest looks good to me for the context of the story.  👍  Pending @cahrens comments being resolved.

cahrens commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -137,6 +140,7 @@ function DragAndDropEditBlock(runtime, element, params) {
                         self.scrollToTop();
+                        $zoneTab.find('input:first').select();
 
                         $(this).one('click', function loadThirdTab(e) {
```
I believe there are preventDefault calls you can remove in this file (and perhaps others) now that the a-tags have been converted to buttons.

itsjeyd replied:
@cahrens Thanks for pointing that out. I removed the `preventDefault` calls for the a-tags that are now buttons. (Taking over for @mtyaka since he is away this week.)

cahrens commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -67,6 +67,9 @@ function DragAndDropEditBlock(runtime, element, params) {
                     _fn.build.$el.feedback.form.find('.problem-mode').trigger('change');
+
+                    // Set focus on first input field.
+                    $element.find('input:first').select();
```
@cptvitamin can weigh in on this during the a11y review. I do wonder which is better-- moving focus to the first input field vs. moving it to the top of the visible text above the field (with tabindex = -1). I suppose it depends on if the text is linked to the input field. For example, on the second "tab", the section "Zones" is not linked to the label/input field and could be missed by a screen reader.

But as I said, that level of detail can wait until the a11y review.

cahrens commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -137,7 +140,7 @@ function DragAndDropEditBlock(runtime, element, params) {
                         $zoneTab.removeClass('hidden');
                         self.scrollToTop();
-                        $zoneTab.find('input:first').focus();
+                        $zoneTab.find('input:first').select();
```
I also don't know which is better-- focus vs. select.

cahrens commented:
👍 It would be good to remove the unneeded event.preventDefault() calls, but I don't need to take a look at this again.

itsjeyd commented:
Rebased https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/88/commits/0af3fc39303aa6bbaf2dc357948b24252e0e9f70 onto master.

itsjeyd commented:
@staubina @cahrens Thanks again for the review. Comments addressed (aside from the one about moving focus that you said could be handled as part of the a11y review, @cahrens).

@cptvitamin @clrux I've added this story to the a11y review ticket for DnDv2.1 ([SOL-1978](https://openedx.atlassian.net/browse/SOL-1978)) so you can have a look at it together with the other DnDv2.1 features. (My bandwidth for working on this PR is limited this week, and since it builds on some DnDv2.1 stories that were already part of SOL-1978, it makes sense to add this PR to the combined review.)

@sstack22 This is ready for you, please have a look.

sstack22 commented:
Looks great, 👍  on my end. Two small nits: capitalization is a bit inconsistent (Final Feedback vs. Item description, as an example).  I would also suggest removing the Note at the top of the Studio Editor - I'm concerned about course teams being prompted to delete problems. As this is outside the scope of this PR, I'm fine with this being addressed separately.

itsjeyd commented:
@sstack22 Thanks for the review! Merging the PR now.

Re: your comments about inconsistent capitalization and removing the "Note", I'll see with @bradenmacdonald whether/how we can incorporate those into the work that's left for the DnDv2.1 epic.
